### PR TITLE
Adding an implementation of Denoised Smoothing

### PR DESCRIPTION
**Pull request by sayakpaul**
This PR adds [Denoised Smoothing](https://arxiv.org/abs/2003.01908) under the `research` folder. 

The success of [Randomized Smoothing](https://arxiv.org/abs/1902.02918) is proven and it works for many different scenarios. But it also operates under the assumption that the underlying classifier is able to perform well under Gaussian perturbations. Won't it be better if we could just take our standard pre-trained image classifiers (including the Cloud APIs) and have the benefits of Randomized Smoothing inside of them in an easy manner? 

That is preciously what Denoised Smoothing does by prepending a Denoiser to an image classifier and still maintains the theoretical guarantees of robustness against L2 attacks. 

Besides, the implementation includes a suite of utilities that may be helpful to generate robustness certificates. I think that gives us a unique opportunity to actually design an API inside NSL that would allow easy generation of robustness certificates. To the best of my knowledge, there does not exist any such framework as we speak.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[submodule "research/denoised_smoothing"]
```
Comment (csferng) [inline .gitmodules:1]: The submodule seems to be deleted in commit 211e728. Please delete this file also if the submodule is not going to be used.

File: research/denoised_smoothing/certification_output/sigma_0.25_denoiser_mse [hunk 2]
```diff
@@ -0,0 +1,501 @@
+idx	label	predict	radius	correct	time
```
Comment (csferng) [inline research/denoised_smoothing/certification_output/sigma_0.25_denoiser_mse:1]: This file (and other files under `certification_output`) are the Colab notebook's output, right? If so, I think we don't need to check in them.

**Review by csferng — CHANGES_REQUESTED** [review]
Thanks for the PR, Sayak. I briefly went through the code and mostly looked fine.

Please run `pylint` to format the code into TensorFlow style (instruction [here](https://www.tensorflow.org/community/contribute/code_style)).

And regarding the `models.tar.gz` archive, maybe push the models to TF Hub instead?

File: research/denoised_smoothing/certification_output/sigma_0.25_denoiser_mse (hunk 2 above)
Comment (sayakpaul) [reply research/denoised_smoothing/certification_output/sigma_0.25_denoiser_mse:1]: Yes. I have kept the outputs for safe housekeeping.

**Comment by sayakpaul** [comment]
@csferng,

> Please run pylint to format the code into TensorFlow style (instruction here).

I have followed most of the instructions I got from the linter barring a few exceptions:

* `scripts/analyze.py` directly comes from [here](https://github.com/microsoft/denoised-smoothing/blob/master/code/analyze.py) without any change (with citation of course). This is why I have kept it intact. 
* `models/resnet20.py` comes from [here](https://git.io/JsPhA) with the license attached.  

Apart from these, I could not resolve the following:

```
scripts/certification_utils.py:57:4: R1705: Unnecessary "else" after "return" (no-else-return)
scripts/certification_utils.py:82:4: R1705: Unnecessary "else" after "return" (no-else-return)

scripts/certify.py:14:0: W9008: Redundant returns documentation (redundant-returns-doc)
scripts/certify.py:30:2: R1732: Consider using 'with' for resource-allocating operations (consider-using-with)
scripts/certify.py:37:2: C0200: Consider using enumerate instead of iterating with range and len (consider-using-enumerate)
scripts/certify.py:57:4: R1732: Consider using 'with' for resource-allocating operations (consider-using-with)

models/dncnn.py:27:13: R1719: The if expression can be replaced with 'bool(test)' (simplifiable-if-expression)
```

Also, I was checking some of the modules under `research` and I felt they do not strictly follow `tensorflow/tools/ci_build/pylintrc` (such as type annotation, trailing newline, etc.).

**Review by csferng — CHANGES_REQUESTED** [review]
Thanks, @sayakpaul.

I suggested a couple changes to address some linter complaints. Others are fine for now. I just have one more ask: 

Given that the code is under the MIT license while the rest of NSL code is under Apache 2.0, please move the directory `research/denoised_smoothing` to `research/third_party/denoised_smoothing`, so that readers can notice the licensing difference. Thanks!

**Comment by sayakpaul** [comment]
@csferng I am happy to change it Apache. Please advice.

**Comment by csferng** [comment]
@sayakpaul, sorry that I didn't make my point clear. Since Google does not own the copyright to those code, please move the directory to `research/third_party/denoised_smoothing`. Either MIT or Apache 2.0 license is fine, and you can choose one you'd like.

**Comment by sayakpaul** [comment]
@csferng done.

**Comment by sayakpaul** [comment]
@csferng could you provide an update on the progress of the PR?

**Review by csferng — COMMENTED** [review]
@sayakpaul The PR looks good. We are finishing the internal review for this PR and will merge afterwards.

**Comment by sayakpaul** [comment]
@csferng just wanted to see if an update is available for this PR.

**Comment by csferng** [comment]
@sayakpaul Sorry for the delay. The PR is merged now.

**Comment by sayakpaul** [comment]
Thank you so much. 

Is it possible to include a note on the `third_party` projects [here](https://github.com/tensorflow/neural-structured-learning/blob/master/research/README.md)?

**Pull request merged**